src/CommunityToolkit/Xamarin.CommunityToolkit/Views/Calendar/Calendar.shared.cs

```diff
@@ -0,0 +1,533 @@
+				var weekOfMonth = date.WeekOfMonth(FirstDayOfWeek);
+
+				UpdateDaysFromPreviousOrNextMonths(weekOfMonth, date, weeksInMonth, daysInMonth);
+
+				if (!ShowWeekends && (date.DayOfWeek is DayOfWeek.Saturday or DayOfWeek.Sunday))
+				{
+					continue;
+				}
+
+				UpdateDay(date, weekOfMonth, isVisible: true);
+			}
+
+			if (SelectedDays?.Count > 0)
+			{
+				SelectDays(SelectedDays);
+			}
+		}
+
+		void UpdateDaysFromPreviousOrNextMonths(
+			int week,
+			DateTime date,
+			int weeksInMonth,
+			int daysInMonth)
+		{
+			if (week is 1
+				&& date.Day is 1
+				&& date.DayOfWeek != FirstDayOfWeek)
+			{
+				var newDate = date;
+
+				do
+				{
+					newDate = newDate.AddDays(-1);
+
+					if (!ShowWeekends && (newDate.DayOfWeek is DayOfWeek.Saturday or DayOfWeek.Sunday))
+					{
+						continue;
+					}
+
+					UpdateDay(newDate, week, isVisible: ShowDaysFromOtherMonths);
+				} while (newDate.DayOfWeek != FirstDayOfWeek);
+			}
+			else if (week == weeksInMonth
+					 && date.Day == daysInMonth
+					 && date.DayOfWeek != FirstDayOfWeek.PreviousOrFirst())
+			{
+				var newDate = date;
+				var lastDayOfWeek = FirstDayOfWeek.PreviousOrFirst();
+
+				do
+				{
+					newDate = newDate.AddDays(1);
+
+					if (!ShowWeekends && (newDate.DayOfWeek is DayOfWeek.Saturday or DayOfWeek.Sunday))
+					{
+						continue;
+					}
+
+					UpdateDay(newDate, week, isVisible: ShowDaysFromOtherMonths);
+				} while (newDate.DayOfWeek != lastDayOfWeek);
+			}
+		}
+
+		void UpdateDay(DateTime date, int week, bool isVisible)
+		{
+			var column = date.DayOfWeek(FirstDayOfWeek, includeWeekends: ShowWeekends) - 1;
+			var row = week - 1;
+
+			var calendarDay = days.FirstOrDefault(x => Grid.GetColumn(x) == column && Grid.GetRow(x) == row);
+
+			if (calendarDay == null)
+			{
+				calendarDay = new CalendarDay
+				{
+					HorizontalOptions = LayoutOptions.CenterAndExpand,
+					VerticalOptions = LayoutOptions.CenterAndExpand,
+					ControlTemplate = DayControlTemplate,
+					Date = date,
+				};
+
+				Grid.SetColumn(calendarDay, column);
+				Grid.SetRow(calendarDay, row);
+
+				var tapGestureRecognizer = new TapGestureRecognizer();
+				tapGestureRecognizer.Tapped += CalendarDay_OnTapped;
+				calendarDay.GestureRecognizers.Add(tapGestureRecognizer);
+
+				days.Add(calendarDay);
+				gridDays.Children.Add(calendarDay);
+			}
+			else
+			{
+				calendarDay.Date = date;
+				calendarDay.ControlTemplate = DayControlTemplate;
+			}
+
+			calendarDay.Opacity = isVisible ? 1 : 0;
+
+			if (isVisible)
+				DayUpdated?.Invoke(this, new CalendarDayUpdatedEventArgs(calendarDay));
+		}
+
+		void CalendarDay_OnTapped(object? sender, EventArgs e)
+		{
+			var calendarDay = (CalendarDay)(sender ?? throw new ArgumentNullException($"{nameof(sender)} cannot be null"));
+
+			if (calendarDay.IsSelectable)
+			{
+				if (SelectionMode is CalendarSelectionMode.SingleSelect)
+				{
+					if (SelectedDays.Any() && SelectedDays.First().Date == calendarDay.Date)
```

Comment:
Let's avoid unnecessary LINQ calls
```suggestion
					if (SelectedDays.Count > 0 && SelectedDays.First().Date == calendarDay.Date)
```